File: leetcode/python/0100_same_tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isSameTree(self, p, q):
        """
        :type p: TreeNode
        :type q: TreeNode
        :rtype: bool
        """
        
        def helper(p, q):
            if p == None and q != None: return False
            if p != None and q == None: return False
            if p != None and q != None:
                if not helper(p.left, q.left): return False
                
                if p.val != q.val: return False

                if not helper(p.right, q.right): return False

            return True

        if not helper(p, q): return False
        return True

File: leetcode/java/0100_same_tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean helper(TreeNode p, TreeNode q) {
        if (p != null && q != null) {
            if (helper(p.left, q.left) == false) return false;
            if (p.val != q.val) return false;
            if (helper(p.right, q.right) == false) return false;
        }

        if (p == null && q != null) return false;
        if (p != null && q == null) return false;
        return true;
    }

    public boolean isSameTree(TreeNode p, TreeNode q) {
        return helper(p, q);
    }
}